### PR TITLE
feat(cli): include eligibility in getReputation CLI method

### PR DESCRIPTION
**Pull request by lrubiorod**

### CODE REVIEW

**Comment by aesedepece** [comment]
image

**Comment by tmpolaczyk** [comment]
This is a breaking change for the JSON-API `getReputation` method, right?

**Comment by lrubiorod** [comment]
> This is a breaking change for the JSON-API `getReputation` method, right?

Yep

File: node/src/actors/chain_manager/handlers.rs [hunk 1]
```diff
@@ -1301,12 +1301,23 @@ impl Handler<GetReputation> for ChainManager {
...
+            is_active,
+        };
+
+        Ok((rep_stats, total_reputation))
     }
 }
 
 impl Handler<GetReputationAll> for ChainManager {
-    type Result = Result<HashMap<PublicKeyHash, (Reputation, bool)>, failure::Error>;
+    type Result = Result<(u64, HashMap<PublicKeyHash, ReputationStats>), failure::Error>;
```
Comment (tmpolaczyk) [inline node/src/actors/chain_manager/handlers.rs:1320]: `GetReputation` returns `(ReputationStats, u64)`, but `GetReputationAll` returns `(u64, HashMap)`. This is a bit confusing, so instead how about explicitly stating the result, like

```rust
struct GetReputationResult {
    stats: HashMap<PublicKeyHash, ReputationStats>,
    total_reputation: u64,
}
```

And we could unify the return types of `GetReputation` and `GetReputationAll`, `GetReputation` would return a `stats` with one element.

**Comment by drcpu-github** [comment]
Is this the number of data requests you were eligible to solve and the number you did actually solve?

**Comment by lrubiorod** [comment]
> Is this the number of data requests you were eligible to solve and the number you did actually solve?

This PR is related with this https://github.com/witnet/witnet-rust/issues/1456.
To avoid reputation oligarchies, we order nodes by reputation and we apply another distribution (trapezoidal distribution). So, if a node has 20000 of reputation and another only 1, it doesnt mean that the first one has a probability 20000 bigger than the other.

Currently in the testnet, your probability to mine a data request is: reputation/total_active_reputation * factor (this factor is related with the number of witnesses required). And with #1456, it will be: eligibility/total_active_reputation * factor, where eligibility depends on your reputation but it is not exactly the same.

And... this PR modifies the json methods for reputation so... you will need to update it in the blockexplorer

File: src/cli/node/json_rpc_client.rs [hunk 2]
```diff
@@ -247,56 +247,53 @@ pub fn get_utxo_info(
...
+                log::info!("Node pkh: {}", node_pkh);
+
+                node_pkh
+            }
         }
     };
 
+    let params = (pkh, all);
     let request = format!(
         r#"{{"jsonrpc": "2.0","method": "getReputation", "params": [{}], "id": "1"}}"#,
```
Comment (tmpolaczyk) [inline src/cli/node/json_rpc_client.rs:277]: Here the request will be formatted as one of the two options, depending on the `--all` flag:

```json
{"jsonrpc": "2.0","method": "getReputation", "params": [["twit1fulan0j78dmdfa3gx0779kszj5sc5j35feljv5",false]], "id": "1"}
{"jsonrpc": "2.0","method": "getReputation", "params": [["twit1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqkz4ay",true]], "id": "1"}
```

As you can see this is pretty strange, I think I prefer the old format, even if it means adding back the `getReputationAll` JSON-RPC method:

```json
{"jsonrpc": "2.0","method": "getReputation", "params": ["twit1fulan0j78dmdfa3gx0779kszj5sc5j35feljv5"], "id": "1"}
{"jsonrpc": "2.0","method": "getReputationAll", "id": "1"}
```

**Comment by drcpu-github** [comment]
Oh, right, I did not link it to the new trapezoidal distribution for solving data requests.

Should not be too big of a change for the explorer, probably just add a bit of extra parsing logic.

File: node/src/actors/messages.rs [hunk 3]
```diff
@@ -253,39 +253,37 @@ impl Message for GetUtxoInfo {
...
+/// GetReputation result
+#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
+pub struct GetReputationResult {
+    /// ReputationsStats hashmap
+    pub stats: HashMap<PublicKeyHash, ReputationStats>,
+    /// Total active reputation
+    pub total_reputation: u64,
+}
+
 /// Get Reputation of one pkh
```
Comment (tmpolaczyk) [inline node/src/actors/messages.rs:276]:
```suggestion
/// Get reputation of one identity if `all` is set to `false`, or all identities if `all` is set to `true`
```

File: node/src/actors/messages.rs [hunk 4]
```diff
@@ -253,39 +253,37 @@ impl Message for GetUtxoInfo {
...
+    /// Eligibility
+    pub eligibility: u32,
+    /// Is active flag
+    pub is_active: bool,
+}
+
+/// GetReputation result
+#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
+pub struct GetReputationResult {
+    /// ReputationsStats hashmap
```
Comment (tmpolaczyk) [inline node/src/actors/messages.rs:270]:
```suggestion
    /// Map of identity public key hash to reputation stats
```

File: node/src/actors/messages.rs [hunk 5]
```diff
@@ -253,39 +253,37 @@ impl Message for GetUtxoInfo {
     type Result = Result<UtxoInfo, failure::Error>;
 }
 
+/// Reputation info
+#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
+pub struct ReputationStats {
+    /// Reputation
+    pub reputation: Reputation,
+    /// Eligibility
```
Comment (tmpolaczyk) [inline node/src/actors/messages.rs:261]: I would be good to add a short explanation on what's the difference between eligibility and reputation, or link to the function that calculates eligibility given reputation.

**Review by tmpolaczyk — APPROVED** [review]
I added some documentation suggestions

**Pull request merged**